api/v1alpha1/stargate_types.go

```diff
@@ -0,0 +1,143 @@
+/*
+Copyright 2021.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// StargateSpec defines the desired state of Stargate
+type StargateSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// DatacenterName is the name of a CassandraDatacenter resource where Stargate should be
+	// deployed.
+	// +kubebuilder:validation:Required
+	DatacenterName string `json:"datacenterName"`
+
+	// Size is the number of Stargate instances to deploy. This value may be scaled independently of
+	// Cassandra cluster nodes. Each instance handles API and coordination tasks for inbound
+	// queries.
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:default=1
+	Size int32 `json:"size"`
+
+	// ServerVersion is the version of Stargate to deploy.
+	// +kubebuilder:validation:Pattern=\d+\.\d+\.\d+
+	// +kubebuilder:default="1.0.29"
+	// FIXME can we use type container.Version or version.Version here?
+	ServerVersion string `json:"serverVersion"`
+
+	// ContainerImage is the image characteristics to use for Stargate containers. Leave nil to use
+	// a default image.
+	// +optional
+	ContainerImage *ContainerImage `json:"containerImage,omitempty"`
+
+	// InitContainerImage is the image characteristics to use for Stargate init containers. Init
+	// containers are used to wait for Cassandra. Leave nil to use a default image.
+	// +optional
+	InitContainerImage *ContainerImage `json:"initContainerImage,omitempty"`
```

Comment:
Keep in mind that a pod can have multiple init containers. If this is for the init container we use in the Helm, then I would include that in the field name. With that said, we won't need that init container :) The purpose of the init container we deploy with the chart is to wait to deploy Stargate until the Cassandra cluster is up. The K8ssandraCluster controller can do this. It will be the one that creates the Stargate and Reaper objects. It should wait to deploy them until the Datacenter is up. Of course, both Stargate and Reaper can be deployed without a K8ssandraCluster, so I would include logic in the Stargate controller to check that the datacenter is ready before deploying Stargate.